### PR TITLE
ENH: add pre_execute dispatcher

### LINKED ISSUE #1189: "pre" execute portions of the Ibis tree
I have a use case where I need to intercept normal Ibis execution and operator on multiple Nodes within the tree all at once.  This operation will result in a partially-filled `scope` dictionary for the relevant Nodes. This then short-circuits the depth-first traversal of the tree, because `scope` has the data it needs to evaluate a given node.  A `pre_execute` dispatcher would allow me to intercept normal execution, do the partial fill of `scope` and return execution control to the regular dispatcher.

### PR DESCRIPTION
**Pull request by toryhaavik**
closes #1189

### CODE REVIEW

**Comment by jreback** [comment]
we should a test for this (and data_preload if we don’t have one there)

eg create a fake client and register these functions and make sure the scope is properly updated

**Comment by cpcloud** [comment]
@toryhaavik Can you add release note in `docs/source/release.rst`?

**Comment by cpcloud** [comment]
Also, can write a small blurb in a comment about why `pre_execute` is useful?

**Comment by toryhaavik** [comment]
Sure, will add docs.  For the comment documenting it, where do you mean?   in the code or in release notes?

**Comment by cpcloud** [comment]
@toryhaavik In the code is fine.

File: ibis/pandas/core.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def find_data(expr):
             seen.add(node)
 
             if hasattr(node, 'source'):
-                data[node] = node.source.dictionary[node.name]
+                data[node] = node.source.dictionary.get(node.name, None)
```
Comment (cpcloud) [inline ibis/pandas/core.py]: Don't need to do this now, but might be worth thinking about whether we can detect if a backend hasn't provided data and provides neither a `pre_execute` nor `data_preload` implementations and raise an error.

**Comment by cpcloud** [comment]
@toryhaavik can you rebase when you get a chance?

**Comment by cpcloud** [comment]
This LGTM.

**Comment by cpcloud** [comment]
Thanks @toryhaavik.

**Pull request closed without merging**